[ROCm][MXFP4] Enable FP4 MLA BMM support

dllehr-amd opened the pull request:
- Added support for FP4 BMM in ROCm aiter operations.
- Introduced environment variable for FP4 BMM configuration.
- Updated quantization utilities to handle dynamic MXFP4 quantization.
- Enhanced MLACommonBaseImpl to utilize FP4 BMM when enabled.

<!-- markdownlint-disable -->

## Purpose

## Test Plan

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

dllehr-amd committed:
[ROCm][MXFP4] Enable FP4 MLA BMM support

- Added support for FP4 BMM in ROCm aiter operations.
- Introduced environment variable for FP4 BMM configuration.
- Updated quantization utilities to handle dynamic MXFP4 quantization.
- Enhanced MLACommonBaseImpl to utilize FP4 BMM when enabled.

Signed-off-by: Douglas Lehr <Doug.Lehr@amd.com>

dllehr-amd committed:
Fix decode_q_nope padding issue

Signed-off-by: Douglas Lehr <Doug.Lehr@amd.com>

tjtanaa commented on vllm/v1/attention/backends/mla/common.py:
```diff
@@ -1237,28 +1238,51 @@ def get_and_maybe_dequant_weights(layer: LinearBase):
+        if self.is_aiter_triton_fp4_bmm_enabled:
+            from aiter.ops.triton.batched_gemm_a16wfp4 import batched_gemm_a16wfp4
+            
+            #print(f'>>> x pre (up_proj) {x.shape}')
```
NITS: let's clean up these

tjtanaa commented on vllm/v1/attention/backends/mla/common.py:
```diff
@@ -1998,7 +2034,26 @@ def forward(
+            if self.is_aiter_triton_fp4_bmm_enabled:
+                from aiter.ops.triton.batched_gemm_a16wfp4 import batched_gemm_a16wfp4
+                
+                #x = x.view(-1, self.num_heads, self.kv_lora_rank)
```
NITS: let's clean up these

tjtanaa commented on vllm/v1/attention/backends/mla/common.py:
```diff
@@ -1998,7 +2034,26 @@ def forward(
+                #x = x.view(-1, self.num_heads, self.kv_lora_rank)
+                x = decode_q_nope.transpose(0, 1)
+                decode_ql_nope = None
+                #print(f'>>> x {x.shape}, q_nope_out {q_nope_out.shape}, self.W_K {self.W_K.shape}')
```
NITS: let's clean up these

tjtanaa commented on vllm/model_executor/layers/quantization/quark/utils.py:
```diff
@@ -103,3 +105,97 @@ def _is_equal_or_regex_match(
+    h, b, d = x.shape
+    x, x_scales = dynamic_mxfp4_quant(x.reshape(-1, d))
+    return x.view(h, b, d // 2), x_scales.view(h, b, d // 32)
+    #return x.view(h, b, d // 2), x_scales.view(h, b, d // 32)
```
NITS: let's clean up these

tjtanaa commented on vllm/v1/attention/backends/mla/common.py:
```diff
@@ -1237,28 +1238,51 @@ def get_and_maybe_dequant_weights(layer: LinearBase):
+                prequant=True,
+                y_scale=None,
             )
+            #print(f'>>> x before transpose {x.shape}')
```
NITS: let's clean up these

tjtanaa commented on vllm/envs.py:
```diff
@@ -117,6 +117,7 @@
     VLLM_ROCM_USE_AITER_FP4_ASM_GEMM: bool = False
     VLLM_ROCM_USE_AITER_TRITON_ROPE: bool = False
     VLLM_ROCM_USE_AITER_FP8BMM: bool = True
+    VLLM_ROCM_USE_AITER_FP4BMM: bool = True
```
Let's infer the conditions based on model weights, model config, and store those evaluated conditions as class properties rather than depending on the flags.

tjtanaa replied:
Please provide the per gain of enabling FP4 MLA BMM support, and also lm_eval score of FP4 MLA BMM and FP8 MLA BMM, since we are consolidating the logic of both FP4 MLA BMM and FP8 MLA BMM in this PR.

tjtanaa commented on vllm/v1/attention/backends/mla/common.py:
```diff
@@ -1998,7 +2034,26 @@ def forward(
 
-            if self.is_aiter_triton_fp8_bmm_enabled:
+            if self.is_aiter_triton_fp4_bmm_enabled:
+                from aiter.ops.triton.batched_gemm_a16wfp4 import batched_gemm_a16wfp4
```
It seems this kernel does not exist in the AITER commit that is currently used in the `docker/Dockerfile.rocm_base` (https://github.com/ROCm/aiter/tree/59bd8ff2c8c3dc1c6caa990a68055528657a1506/aiter/ops/triton). So, we will only be able to merge after we upgrade the AITER version.

dllehr-amd committed:
Remove extraneous print and comments

Signed-off-by: Douglas Lehr <Doug.Lehr@amd.com>

dllehr-amd committed:
Remove env var and automatically detect when to use mxfp4 bmms

Quark's DSFP4 will have unquantized projection weights.  We can use that to
detect if we need to quantize down to mxfp4.  Normally Deepseek will have the
projection weights as fp8.

Signed-off-by: Douglas Lehr <Doug.Lehr@amd.com>

dllehr-amd committed:
Ruff Cleanup

Signed-off-by: Douglas Lehr <Doug.Lehr@amd.com>
